sbt-scalajs-bundler/src/main/scala/scalajsbundler/sbtplugin/NpmDepsPlugin.scala

```diff
@@ -0,0 +1,152 @@
+package scalajsbundler.sbtplugin
+
+import org.scalajs.core.tools.io.FileVirtualJSFile
+import org.scalajs.core.tools.jsdep.ResolvedJSDependency
+import org.scalajs.sbtplugin.ScalaJSPlugin
+import org.scalajs.sbtplugin.ScalaJSPlugin.autoImport._
+import sbt.Keys._
+import sbt._
+
+import scalajsbundler.NpmDeps.NpmDeps
+import scalajsbundler._
+
+object NpmDepsPlugin extends AutoPlugin {
+
+  override lazy val requires = ScalaJSPlugin
+
+  // Exported keys
+  object autoImport {
+
+    // From ScalaJSBundlerPlugin
+    val npmUpdate = taskKey[File]("Fetch NPM dependencies")
+
+    // From ScalaJSBundlerPlugin
+    val useYarn: SettingKey[Boolean] =
+      settingKey[Boolean]("Whether to use yarn for updates")
+
+    // From ScalaJSBundlerPlugin
+    val packageJson = TaskKey[BundlerFile.PackageJson]("packageJson",
+      "Write a package.json file defining the NPM dependencies of project",
+      KeyRanks.Invisible
+    )
+
+    val npmDeps = settingKey[NpmDeps]("List of js dependencies to be fetched")
+
+    val allNpmDeps = taskKey[NpmDeps]("json file containing all npm js dependencies collected from all dependencies")
+
+    val dependencyFile = taskKey[File]("File containing all external js files")
+  }
+
+  import autoImport._
+
+  override lazy val projectSettings = Seq(
+    npmDeps in Compile := List.empty,
+
+    skip in packageJSDependencies := false,
+
+    resolvedJSDependencies in Compile := {
+      val logger = streams.value.log
+      val prev = (resolvedJSDependencies in Compile).value
+
+      // Fetch the js paths in node_modules
+      val jss = {
+        val nodeModules = (npmUpdate in Compile).value / "node_modules"
+
+        (for {
+          m <- (allNpmDeps in Compile).value
+          js <- m.jsFiles
+        } yield {
+          logger.info(s"Fetch $js in ${nodeModules / m.module}")
+          get(nodeModules / m.module, js)
+        }).flatten
+
+      }
+
+      val resolvedDependencies = jss.map { f =>
+        ResolvedJSDependency.minimal(FileVirtualJSFile(f))
+      }
+
+      prev.map(_ ++ resolvedDependencies)
+    },
+
+    dependencyFile := (packageMinifiedJSDependencies in Compile).value,
```

Comment:
This is defined, but never *read*. Looks like dead code.

Reply:
dependencyFile is the Key to get the packagedMinifiedJSDependencies. I use it in my demo to build deps.js.

Reply:
Why don't you use `(packageMinifiedJSDependencies in Compile).value` in your demo instead, then?

Reply:
To render the API smarter. Here, you only have to call ```dependencyFile.value```, and not both:
```scala
skip in packageJSDependencies := false
...
packageMinifiedJSDependencies in Compile).value
```

Reply:
In the current definition of the plugin, `dependencyFile` does not influence `skip in packageJSDependencies`, so anyway you need two "things":
```scala
skip in packageJSDependencies := false
...
dependencyFile.file
```
I do not see the added value.

Reply:
Yes, but it is hidden in the NpmDepsPlugin project settings. So that, you only have one line to write in your project definition and not 2.

Reply:
My comment here does not suggest removing `skip in packageJSDependencies := false` from the plugin. If you just remove the `dependencyFile` from the sbt plugin, then the only thing that changes in the demo is to replace `dependencyFile.value` by `(packageMinifiedJSDependencies in Compile).value`. It is still one line in the project definition.

Reply:
You're right. On the other hand, isn't it a little bit confusing since these two lines are still use together in scalajs ?

Reply:
> On the other hand, isn't it a little bit confusing since these two lines are still use together in scalajs ?

The whole settings around `jsDependencies` is confusing because it's mostly broken, and that's why it's been extracted away so that people who don't need it can ignore it.

Reply:
No. I recommend that `NpmDepsPlugins` only takes care of `npmDependencies`, `npmUpdate`, and immediate friends that do not depend on `jsDependencies` (e.g., `useYarn`, `packageJson`).

Then, you could start a new project that links `NpmDepsPlugins` with `jsDependencies`, `resolvedJSDependencies`, and friends (e.g., `packageMinifiedJSDependencies).

`NpmDepsPlugin` would be the common denominator of `ScalaJSBundlerPlugin` and your other plugin.

Less magic, less responsibilities per plugin -> win/win for everyone, because everyone can choose how much they need.

Reply:
I think it should be a very different project, so that the scalajs-bundler repo can eventually be rid of the dependency on (what's been moved to) sbt-jsdependencies.

That said, you can probably *develop* it first in the same repo so that you can iterate more quickly, but then PR only the part for `NpmDepsPlugin`.